mmeval/metrics/coco_panoptic.py

```diff
@@ -0,0 +1,656 @@
+            if sem_label == len(self.classes):  # type: ignore # yapf: disable # noqa: E501
+                continue
+            mask = pan == pan_label
+            area = mask.sum()
+            # when ann_file provided, sem_label should be cat_id, otherwise
+            # sem_label should be a continuous id, not the cat_id
+            # defined in dataset
+            category_id = self.label2cat[sem_label] if self.label2cat else sem_label   # type: ignore # yapf: disable # noqa: E501
+            segments_info.append({
+                'id': int(pan_label),
+                'category_id': category_id,
+                'area': int(area)
+            })
+
+        # evaluation script uses 0 for VOID label.
+        pan[pan % INSTANCE_OFFSET == len(self.classes)] = VOID  # type: ignore # yapf: disable # noqa: E501
+        pan = id2rgb(pan).astype(np.uint8)
+
+        segm_file = osp.join(self.outfile_prefix, prediction['segm_file'])  # type: ignore # yapf: disable # noqa: E501
+        if HAS_MMCV:
+            mmcv.imwrite(pan[:, :, ::-1], segm_file)
+        else:
+            imwrite(pan[:, :, ::-1], segm_file)
+
+        prediction['segments_info'] = segments_info
+        return prediction
+
+    @staticmethod
+    def _compute_pq_stats(prediction: dict,
+                          groundtruth: dict,
+                          categories: dict,
+                          seg_prefix: str,
+                          outfile_prefix: str,
+                          backend_args: Optional[dict] = None,
+                          coco_api: Optional[PanopticAPI] = None) -> PQStat:
+        """The function to compute the metric of Panoptic Segmentation.
+
+        Args:
+            prediction (dict): Same as :class:`CocoPanoptic.add`.
+            groundtruth (dict): Same as :class:`CocoPanoptic.add`.
+            categories (dict): The categories of the dataset.
+            seg_prefix (str): Same as :class:`CocoPanoptic.seg_prefix`.
+            outfile_prefix (str): Same as :class:`CocoPanoptic.outfile_prefix`.
+            backend_args (dict, optional): Same as
+                :class:`CocoPanoptic.backend_args`.
+            coco_api (PanopticAPI, optional): PanopticAPI wrapper.
+                Defaults to None.
+
+        Returns:
+            PQStat: The metric results of Panoptic Segmentation.
+        """
+        # get panoptic groundtruth and prediction
+        gt_seg_map_path = osp.join(seg_prefix, groundtruth['segm_file'])
+        pred_seg_map_path = osp.join(outfile_prefix, prediction['segm_file'])
+
+        if HAS_MMCV:
+            img_bytes = get(gt_seg_map_path, backend_args=backend_args)
+            pan_gt = mmcv.imfrombytes(
+                img_bytes, flag='color', channel_order='rgb')
+            img_bytes = get(pred_seg_map_path, backend_args=backend_args)
+            pan_pred = mmcv.imfrombytes(
+                img_bytes, flag='color', channel_order='rgb')
+        else:
+            pan_gt = imread(
```

Comment:
If no mmcv, does it not support loading from other file clients?

Reply:
`imread` in mmeval is a simplified version of mmcv's. If we want to support loading from other file clients, yes, it is necessary. Otherwise, we can simple use cv2 or PIL to read the image, but it is not fast enough. (imread and imwrite still need in CityScapesMetric)